pkg/datastore/test/caveat.go

```diff
@@ -81,29 +99,62 @@ func WriteCaveatedRelationshipTest(t *testing.T, tester DatastoreTester) {
 
 	// Store caveat, write caveated tuple and read back same value
 	coreCaveat := createCoreCaveat(t)
+	anotherCoreCaveat := createCoreCaveat(t)
 	ctx := context.Background()
-	_, err = writeCaveat(ctx, ds, coreCaveat)
+	_, err = writeCaveats(ctx, ds, coreCaveat, anotherCoreCaveat)
 	req.NoError(err)
 
 	tpl := createTestCaveatedTuple(t, "document:companyplan#parent@folder:company#...", coreCaveat.Name)
 	rev, err := common.WriteTuples(ctx, sds, core.RelationTupleUpdate_CREATE, tpl)
 	req.NoError(err)
-	iter, err := ds.SnapshotReader(rev).QueryRelationships(ctx, datastore.RelationshipsFilter{
-		ResourceType: tpl.ResourceAndRelation.Namespace,
-	})
+	assertTupleCorrectlyStored(req, ds, rev, tpl)
+
+	// RelationTupleUpdate_CREATE of the same tuple and different caveat context will fail
+	_, err = common.WriteTuples(ctx, sds, core.RelationTupleUpdate_CREATE, tpl)
+	req.ErrorAs(err, &common.CreateRelationshipExistsError{})
+
+	// RelationTupleUpdate_TOUCH does update the caveat context for a caveated relationship that already exists
+	currentMap := tpl.Caveat.Context.AsMap()
+	delete(currentMap, "b")
+	st, err := structpb.NewStruct(currentMap)
+	require.NoError(t, err)
+
+	tpl.Caveat.Context = st
+	rev, err = common.WriteTuples(ctx, sds, core.RelationTupleUpdate_TOUCH, tpl)
 	req.NoError(err)
+	assertTupleCorrectlyStored(req, ds, rev, tpl)
 
-	defer iter.Close()
-	readTpl := iter.Next()
-	foundDiff := cmp.Diff(tpl, readTpl, protocmp.Transform())
-	req.Empty(foundDiff)
+	// TOUCH can remove caveat from relationship
+	caveatContext := tpl.Caveat
+	tpl.Caveat = nil
+	rev, err = common.WriteTuples(ctx, sds, core.RelationTupleUpdate_TOUCH, tpl)
+	req.NoError(err)
+	assertTupleCorrectlyStored(req, ds, rev, tpl)
+
+	// TOUCH can store caveat in relationship with no caveat
+	tpl.Caveat = caveatContext
+	rev, err = common.WriteTuples(ctx, sds, core.RelationTupleUpdate_TOUCH, tpl)
+	req.NoError(err)
+	assertTupleCorrectlyStored(req, ds, rev, tpl)
 
 	// Caveated tuple can reference non-existing caveat - controller layer is responsible for validation
```

Comment:
Just thought of one more test for here: write a caveated relationship and ensure it can be deleted without the caveat?